### PR TITLE
Adding Path.normalize() method

### PR DESCRIPTION
**Pull request by linclelinkpart5**
As per the discussion here: https://github.com/rust-lang/rfcs/issues/2208

I found myself with a need to normalize file paths. By "normalize", I refer to the lexical cleanup of a path, following a similar algorithm as Python's [os.path.normpath()](https://docs.python.org/3/library/os.path.html#os.path.normpath) and Go's [filepath.Clean()](https://golang.org/pkg/path/filepath/#Clean). These in turn follow the methodology outlined in Rob Pike's paper: [Lexical File Names in Plan 9](https://9p.io/sys/doc/lexnames.html).

I was able to write a method that works for my use case, and I felt it could be a helpful addition to stdlib! However, I am very new to both Rust and contributing to OSS projects. Please do let me know if there are any changes that could/should be made in those regards.

### CODE REVIEW

**Comment by retep998** [comment]
I don't see any tests for `\\.\` device paths or `\foo\bar` rooted paths.

**Comment by ThatsGobbles** [comment]
@retep998: Those are good points, I can add those tests. What would be some other good Windows-specific tests to include?

I must note, I don't have access to a Windows dev box, so it's tricky to run these Windows-specific tests.

**Comment by retep998** [comment]
At the very least, we should have every single example from the tables on https://googleprojectzero.blogspot.com/2016/02/the-definitive-guide-on-win32-to-nt.html

**Comment by joshtriplett** [comment]
You should also include test cases for filenames containing spaces, tabs, newlines, and all three.

Please use raw strings for filenames containing backslashes, so that the test cases don't need to escape backslashes.

**Comment by retep998** [comment]
@rust-lang/libs Do we want to use `GetFullPathNameW` on Windows or do we want to roll our own normalization, and if we do choose the latter, where exactly should we differ from the behavior of `GetFullPathNameW`?

**Comment by joshtriplett** [comment]
@retep998 If we *don't* use `GetFullPathNameW` on Windows, it seems worth using it as part of the test cases, by calling both it and `Path.normalize()` on every test case and failing if they differ. (Unless we have some reason to differ, in which case we should have specific test cases for that.)

**Comment by ThatsGobbles** [comment]
Does `GetFullPathNameW` use the underlying filesystem? My goal with this method was to perform purely lexical conversion of a path.

**Comment by joshtriplett** [comment]
@ThatsGobbles According to the MSDN documentation, "This function does not verify that the resulting path and file name are valid, or that they see an existing file on the associated volume."

It does, however, look at the process's current working directory.

**Comment by ThatsGobbles** [comment]
I managed to get a Windows 10 VM up and running, and am trying out tests. The normalization does not seem to like the examples in The Definitive Guide to NT and Win32 Paths involving trailing dots and spaces at the end of a path (the `DEF. .` examples), or using COM ports. It seems that `Path.components()` isn't treating these as special cases?

**Comment by ThatsGobbles** [comment]
It appears that [this Windows shell function](https://msdn.microsoft.com/en-us/library/windows/desktop/hh707084(v=vs.85).aspx) does what I'm aiming to do with my function!

Now the question would be, how do I get access to it in my Rust code?

**Comment by ThatsGobbles** [comment]
As it turns out, I'm not really sure what I should be doing to get this working. I don't know anything about Windows C FFI in Rust, which is stopping me from using the PathCch methods (not to mention, I'd like for normalize to be a pure function that can't fail, and relying on a Windows API seems like it introduces the possibility of failure).

Are there any guides or reading I can look at to get my feet wet with this?

**Comment by joshtriplett** [comment]
I'm going to defer to @retep998 here.

**Comment by carols10cents** [comment]
Ping @retep998! Do you have any advice for @ThatsGobbles ?

**Comment by retep998** [comment]
* Is the intent of this function to turn relative paths into absolute paths or is it simply trying to clean up `..` and `.` and slash direction?
* If I recall correctly, on linux `foo/bar/..` may not necessarily be `foo` because `bar` could be a symbolic link resulting in a completely different directory if accessed directly versus accessing the normalized version. Is this inaccuracy an acceptable tradeoff?

In my opinion, we need to rethink what transformations are actually useful. Here are a few legitimate scenarios I thought of. If you can think of any more, feel free to bring them up.

* The user has a relative path and wants to turn it into an absolute path so that it no longer needs to rely on the current directory. Naively joining the path onto the current directory won't work on Windows (due to per drive current directories) so there needs to be a function that can do this reliably in a cross platform manner. While `fs::canonicalize` can do this, it unnecessarily hits the filesystem in order to resolve symbolic links and also requires that the path actually exists which is overkill when all you want is an absolute path.
* The user has two paths and wants to know whether they point to the same file. Canonicalization is an invalid solution because a file that has hard links technically has multiple canonical paths. The correct solution is to get the IDs of both files and compare that. Currently the `same-file` crate can be used to do this.
* The user has a path with `..` and `.` in it or perhaps mixed slash directions on Windows, and wants to clean it up. On linux due to the way `..` and `.` are symbolic links, this would have to call `realpath` in order to not get a different result than accessing the original path, and would have to hit the filesystem. On Windows this can be done fairly easily without touching the filesystem as long as careful attention is paid to the prefix of the path to not alter them in `\\?\` paths.

**Comment by ExpHP** [comment]
My thoughts, as someone who has been looking for a function like this:

> Is the intent of this function to turn relative paths into absolute paths

Definitely not.  In fact, preservation of leading `"../"` should be considered an explicit goal, and I'm glad that such an example is included in the documentation.

> If I recall correctly, on linux `/foo/bar/..` may not necessarily be `/foo` because `bar` could be a symbolic link resulting in a completely different directory if accessing the original path versus accessing the normalized version. Is this inaccuracy an acceptable tradeoff?

If you ask me, **that's the entire point.**

**Technically,** `.` and `..` in linux are symbolic links pointing to the parent, and if you were to interpret the path `/a/b/c/../..` as a traversal through the directory graph, you could very well end up somewhere other than `/a`.

**In practice,** *nobody enjoys that mental model very much at all!*  Lots of linux software deliberately uses methods like `normalize` to hide the ugly nature of the directory graph and present a mental model of the directory structure that is more like an infinitely expanding tree, where `..` simply means "back the way you came."

Personally, I have yet to see a shell where `cd /a/b/c/../..` will send you anywhere other than `/a`.

**Comment by ThatsGobbles** [comment]
Thank you for the helpful pings and chiming in!

To answer your bullet points, @retep998:
* The goal of this function is simply to clean up dot references (`.` and `..`), and normalize slashes (`/` on POSIX and `\` on Windows, verbatim paths aside).
* Yes, this would be fine (and expected) in the use case I have in mind.

And, in terms of the scenarios:
* This would be a useful method, as a quick way to make a path absolute. However, the `normalize` method I have in mind would leave relative paths as-is. Some POSIX examples:
```
# curdirs, trailing slashes, and multiple consecutive slashes are collapsed
././wow////nice/ -> wow/nice

# if the result is empty return a single dot
./ -> .

# updirs consume the previous component...
ugh/../cool/ -> cool

# ...but are left prepended if at the start of a relative path
../../../neat -> ../../../neat
```
The 'absolute-tize' function described here would be closest to Python's [os.path.abspath](https://docs.python.org/3/library/os.path.html#os.path.abspath), while the functionality I have in mind for normalization would be closest to Python's [os.path.normpath](https://docs.python.org/3/library/os.path.html#os.path.normpath). To complete the analogies, Rust's existing `fs::canonicalize` would be closest to Python's [os.path.realpath](https://docs.python.org/3/library/os.path.html#os.path.realpath).
* The same file example I feel is best left to the aformentioned crate, this would not try and be a replacement.

**Comment by ExpHP** [comment]
By the way, I really like Go's name **`clean`**, which I feel very clearly states the purpose and is less easily confused with `canonicalize`.  (I also notice that it was called "cleanpath" in the Plan 9 paper.  I wonder why they decided to call it normpath in python?)

**Comment by kennytm** [comment]
Hi @retep998, OP has responded to you at https://github.com/rust-lang/rust/pull/47363#issuecomment-359978077

Also, @ThatsGobbles if you're still working on this, you need to give the `into_rc()` function back its `#[test]` attribute before the CI check will succeed.

<details>

```
[01:12:57]    Compiling alloc v0.0.0 (file:///checkout/src/liballoc)
[01:13:47] error: function is never used: `into_rc`
[01:13:47]     --> libstd/path.rs:4184:5
[01:13:47]      |
[01:13:47] 4184 |     fn into_rc() {
[01:13:47]      |     ^^^^^^^^^^^^
[01:13:47]      |
[01:13:47] note: lint level defined here
[01:13:47]     --> libstd/lib.rs:232:31
[01:13:47]      |
[01:13:47] 232  | #![cfg_attr(not(stage0), deny(warnings))]
[01:13:47]      |                               ^^^^^^^^
[01:13:47]      = note: #[deny(dead_code)] implied by #[deny(warnings)]
```

</details>

**Comment by retep998** [comment]
I'm definitely prefer if this used a name more in line with the fact that is simply cleaning the path. It is not performing full normalization, so it should **not** be called `normalize`. I do think there needs to be a `normalize` function in std though, but this function is not that.

**Comment by joshtriplett** [comment]
:+1: for `clean` or similar.

**Comment by pietroalbini** [comment]
@retep998 and @ThatsGobbles, ping from triage!

**Comment by retep998** [comment]
Currently what this PR needs to do is:

1. Rename the function to match what it is actually trying to do, which is merely cleaning up `..` and `.` and slash direction. It is *not* trying to turn a relative path into an absolute path, so it should be named something like `clean`, **not** `normalize`.
2. Expand the test suite. The blog post I linked earlier describes all the different types of windows paths, and the test suite should have multiple test cases for each path type to ensure correct behavior.

**Comment by ThatsGobbles** [comment]
Hello all,

I concur with the name change: `clean` is far more clear as to what the method is trying to do!

As for tests, I've been incorporating the examples from the linked page, but I'm running into issues with Windows paths. It seems that `Path.components()` trips up some of the edge cases, and since `clean` (formerly known as `normalize`) depends heavily on `components`, I'm not seeing correct results. Perhaps I'm missing something, either code- or understanding-wise?

File: src/libstd/path.rs
Comment (kennytm) [inline src/libstd/path.rs]: Feature names usually don't contain spaces.

**Review by kennytm — CHANGES_REQUESTED** [review]

**Review by kennytm — APPROVED** [review]

File: src/libstd/path.rs
Comment (kennytm) [inline src/libstd/path.rs]: Nit pick. Unless your path contains a `"` you could just lose the `#`s

```rust
tc!(r"X:\ABC\DEF", r"X:\ABC\DEF");
```

I think this is much less noisy.

**Comment by BatmanAoD** [comment]
Ping from triage! @kennytm and @joshtriplett , it looks like the author needs some help dealing with Windows paths. Can either of you recommend someone who might be able to help? Or should we tag a team? I've also asked on IRC.

**Comment by kennytm** [comment]
@ThatsGobbles Could you lists the test cases which tripped up `components()`?

**Comment by ThatsGobbles** [comment]
Hi @kennytm and @ArazAbishov, apologies for the slow response, I found myself with little free time this week! Here's the current set of problem cases that I've discovered:

```
X:\ABC\DEF. .
X:DEF. .
\ABC\DEF. .
ABC\DEF. .
\\server\share\ABC. .
//server/share/ABC/DEF
\\.\X:\ABC\DEF. .
\\.\X:/ABC/DEF
\\.\X:\ABC\..\..\C:\
COM1
X:\COM1
X:COM1
valid\COM1
X:\notvalid\COM1
X:\COM1.blah
X:\COM1:blah
X:\COM1  .blah
\\.\X:\COM1
\\abc\xyz\COM1
\\?\X:\ABC\DEF. .
\\?\X:/ABC/DEF
\??\X:\ABC\DEF
\??\X:\
\??\X:
\??\X:\COM1
\??\X:\ABC\DEF. .
\??\X:/ABC/DEF
\??\X:\ABC\..\XYZ
\??\X:\ABC\..\..\..
```
All of the cases ending in '. .' failed due to those not getting normalized away. Most of the UNC paths did not trip up `.components()` per say, but when re-joining, there would be an unexpected trailing slash.

**Comment by pietroalbini** [comment]
Ping from the release team! @kennytm the author seems to be still stuck, but replied to your request for information. Can you (or @retep998) help them out?

**Comment by kennytm** [comment]
@ThatsGobbles Hi sorry for long delay. I'm not totally familiar with Windows path, so perhaps @retep998 can give some more mentoring. (The failing test cases are in the source file already)

1. **Cases ending with `. .`**. 

    I think this `components()` should still return `"DEF. ."` as is, meaning the removal of the dots and spaces should be done inside the `clean()` function. You may want to add a `sys::path::clean_file_name` function which will remove `.` and ` ` from the last path component.

    ```rust
    // Pseudo-code. OsStr::trim_right does not exist, implement it yourself.
    #[cfg(windows)]
    pub fn clean_file_name(f: &OsStr) -> &OsStr {
        f.trim_right(&['.', ' '])
    }
    #[cfg(not(windows))]
    pub fn clean_file_name(f: &OsStr) -> &OsStr {
        f
    }
    ```

    You also need to add a test to ensure trailing dots and slashes in the middle are not trimmed out.

    ```rust
    tc!("X:\ABC...\DEF...", "X:\ABC...\DEF");
    tc!("X:\ABC...\DEF...\", "X:\ABC...\DEF...\");
    ```

    **The `components()` iterator will not recognize the trailing slash.** You might need to manually check `self.inner.ends_with(MAIN_SEPARATOR)`.

2. **Prefix with `/` i.e. `//server/share/ABC/DEF`**. 

    The Project Zero blog suggests replacing all `/` with `\` as the first step, which we will not be doing due to immutability 🙂. However, we may modify [`sys::path::parse_prefix`](https://github.com/rust-lang/rust/blob/94ede93467beb4ac17956845607a7e8b226dda02/src/libstd/sys/windows/path.rs#L32) to recognize `/` as the same as `\`. 

    We need to add an argument, `fn parse_prefix(path: &OsStr, is_cleaning: bool) -> Option<Prefix>` to restrict the new behavior within the `clean()` function. This means you have to construct the `Components` instance yourself, instead of calling the public `self.components()`

3. **UNC with middle `/`**

    Instead of always pushing `item.as_os_str()`, you may need to special-case it for UNC prefixes, so that the prefix `\\server\share` would be pushed separately as two components. This allows us to construct a `\\server\share` even if the source was `\\server/share`.

4. **`COM1`**. 

    You'll need to hard-code the reserved filenames. 

    ```rust
    #[cfg(windows)]
    fn clean_reserved_path(filename: &OsStr) -> Option<PathBuf> {
        if let Some(capture) = filename.capture(r"^(COM[0-9¹²³]|…)(?:[:. ]*[:.].*)?$") {
           Some(r"\\.\" + capture.get(1))
        } else {
           None
        }
    }
    #[cfg(not(windows))]
    fn clean_reserved_path(_: &OsStr) -> Option<PathBuf> { 
        None
    }
    ```

    In your `clean()`, if you find the prefix is Drive Absolute, Drive Relative or Relative, you need to call `clean_reserved_path()` on the filename and return it directly if it is `Some`.

    ```rust
    match prefix {
        None | Some(Disk(_)) => {
            if let Some(cleaned) = sys::path::clean_reserved_path(file_name) {
                return cleaned;
            }
        }
        _ => {}
    }
    ```

5. **Trailing slashes**

    I think normalizing `\\foo` into `\\foo\` and `\\foo\bar` into `\\foo\bar\` is OK.

**Comment by shepmaster** [comment]
Ping from triage, @ThatsGobbles ! You've got some feedback from @kennytm; will you be able to address it in the near future?

**Comment by pietroalbini** [comment]
Thank you so much for this PR @ThatsGobbles! Unfortunately, we haven't heard from you in a while, so I'm closing this to keep the list of PRs clean.

If you have time to work on this again please reopen the pull request though! We will be happy to review the new changes.

**Pull request closed without merging**
